### ISSUE TITLE
Have a 'fill-or-kill' for a whole transaction and not just an operation

### ISSUE THREAD

**Issue by xeroc** [cryptonomex/graphene#609]
In order to allow atomic trading from the client-side, we can simply put a set of orders into a transaction (in the right order) and have them executed atomically in the single-threaded engine.

The only issue is that someone else could snatch an order in an intermediate market just before you ops are executed. That would leave the trader in an intermediate asset.

To fix this, we could put a "fill-or-kill" like option to the whole transaction constituting several order operations .. If one of them does not get filled instantly (no partial fill), non of them will get executed.

The question would be what happens if in an intermediate asset you end up with more shares than you asked for (because of a new sell order). Could we have those 'extra' be forwarded as well?

Alternatively, the trader could just ask for a trade across several markets with sell_amount and minimum_to_receive amounts where the trader has to define the set of markets and that order would only be executed or gets killed.

Thoughts?

**Comment by abitmore** [comment]
This can be done via temp-account, as discussed in forum.

However, would this feature lead to some kind of spamming/DDoS attacks? Since all operations will be rolled back, if I understood correctly, all fees will be refunded? So if someone craft a large transaction which contains many expensive operations, but fail in the end, will it consume a lot of resources of the witness_nodes but cost the attacker nothing?

**Comment by xeroc** [comment]
> However, would this feature lead to some kind of spamming/DDoS attacks? Since all operations will be rolled back, if I understood correctly, all fees will be refunded? So if someone craft a large transaction which contains many expensive operations, but fail in the end, will it consume a lot of resources of the witness_nodes but cost the attacker nothing?

If that became an issue we could let the block producers limit the amount of processing available per tmp-account. That would not even require a hard fork but only a simple config parameter. We could limit 
- number of operations
- size of the transaction

and could do so on a
- per account-basis
- per stake basis
  
  or something similar

**Comment by abitmore** [comment]
I have an idea: anytime when _the temp-account is used(every operation)_ or maybe _multiple operations in a transaction_, the initial user must pay a fee according to number of successful operations, the fee won't be refunded no matter the whole transaction is successful or failed in the end.  Thoughts?

**Comment by theoreticalbts** [comment]
> if someone craft a large transaction which contains many expensive operations, but fail in the end, will it consume a lot of resources of the witness_nodes but cost the attacker nothing?

No, the entire p2p network basically acts as a filter on this kind of transaction.  Every node ensures such a transaction is not propagated.  The design of operations ensures each operation requires a bounded amount of work, and overall size limit on transactions ensures the transaction as a whole requires a bounded amount of work.

**Comment by abitmore** [comment]
> > if someone craft a large transaction which contains many expensive operations, but fail in the end, will it consume a lot of resources of the witness_nodes but cost the attacker nothing?
> 
> No, the entire p2p network basically acts as a filter on this kind of transaction. Every node ensures such a transaction is not propagated. The design of operations ensures each operation requires a bounded amount of work, and overall size limit on transactions ensures the transaction as a whole requires a bounded amount of work.

Sorry I didn't understand. I mean if the transaction itself is valid, so the p2p network will relay the transaction normally; but the last operation will fail, for example selling of 100M BTS which the owner doesn't have, won't this be evaluated only when pushing to a block?

**Comment by xeroc** [comment]
> Sorry I didn't understand. I mean if the transaction itself is valid, so the p2p network will relay the transaction normally; but the last operation will fail, for example selling of 100M BTS which the owner doesn't have, won't this be evaluated only when pushing to a block?

From what I understood, this is only considered an attack if it requires more resources than the node wants to provide. In that case, the node won't propagate the transaction and it might not even reach a block producer

**Comment by abitmore** [comment]
> > Sorry I didn't understand. I mean if the transaction itself is valid, so the p2p network will relay the transaction normally; but the last operation will fail, for example selling of 100M BTS which the owner doesn't have, won't this be evaluated only when pushing to a block?
> 
> From what I understood, this is only considered an attack if it requires more resources than the node wants to provide. In that case, the node won't propagate the transaction and it might not even reach a block producer

Imo the validation work of the transaction itself is relatively small, so the nodes WILL propagate it. But the execution of the operations in the transaction could be costly. Thoughts?

**Comment by pmconrad** [comment]
Do we currently have costly operations?
It seems we have, asset_global_settle is potentially costly because it iterates over all shorts of the given asset.
And AFAIR the _blind_ operations are computationally expensive.

I'm not sure how the p2p code and the block production code work together. This should be checked, and it should be made sure that transactions are propagated only after they've successfully been included in a "proto-block". But that should go into another ticket, and this here should be closed.

**Comment by vikramrajkumar** [comment]
This issue was moved to bitshares/bitshares-core#189
